datajunction-server/datajunction_server/api/nodes.py

```diff
@@ -709,6 +712,133 @@ def link_dimension(  # pylint: disable=too-many-arguments
+    If a link already exists, updates the link definition.
+    """
+    node = get_node_by_name(session=session, name=name)
+    if node.type not in (NodeType.SOURCE, NodeType.DIMENSION, NodeType.TRANSFORM):
+        raise DJInvalidInputException(
+            message=f"Cannot link dimension to a node of type {node.type}. "
+            "Must be a source, dimension, or transform node.",
+        )
+
+    # Find the dimension node and check that the catalogs match
+    dimension_node = get_node_by_name(
+        session=session,
+        name=link_input.dimension_node,
+        node_type=NodeType.DIMENSION,
+    )
+    if (
+        dimension_node.current.catalog_id != UNKNOWN_CATALOG_ID
+        and dimension_node.current.catalog is not None
+        and node.current.catalog.name != dimension_node.current.catalog.name
+    ):
+        raise DJException(  # pragma: no cover
+            message=(
+                "Cannot link dimension to node, because catalogs do not match: "
+                f"{node.current.catalog.name}, {dimension_node.current.catalog.name}"
+            ),
+        )
+
+    # Parse the join query and do some basic verification of its validity
+    join_query = parse(f"SELECT 1 FROM {link_input.join_sql}")
+    exc = DJException()
+    ctx = ast.CompileContext(session=session, exception=exc)
+    join_query.compile(ctx)
+
+    # Check that there is a valid join clause
+    if not join_query.select.from_.relations[0].extensions:  # type: ignore
+        raise DJInvalidInputException(
+            f"Provided SQL `{link_input.join_sql}` does contain JOIN clause",
+        )
+    join_relation = join_query.select.from_.relations[0].extensions[0]  # type: ignore
+
+    # Verify that the query references both the node and the dimension being joined
+    expected_references = {name, link_input.dimension_node}
+    references = {table.name.identifier() for table in join_query.find_all(ast.Table)}
+    if expected_references.difference(references):
+        raise DJInvalidInputException(
+            message=f"The join SQL provided does not reference both the origin node {name} and the "
+            f"dimension node {link_input.dimension_node} that it's being joined to.",
+        )
+
+    # Verify that the columns in the ON clause exist on both nodes
+    for col in join_query.find_all(ast.Column):
+        if not col.table or not col.table.dj_node:  # type: ignore
+            raise DJException(
+                message=f"Column {col.identifier()} does not exist on node",
+                http_status_code=404,
+            )
+
+    # Find an existing dimension link if there is already one defined for this node
+    existing_link = [
+        link
+        for link in node.current.dimension_links
+        if link.dimension_id == dimension_node.id
+    ]
+    is_update = False
+
+    if existing_link:
+        # Update the existing dimension link
+        is_update = True
+        dimension_link = existing_link[0]
```

Comment:
I think this `...[0]` only makes sense if we assume (and validate somewhere) that a node can only be joined to a dimension once. But in reality this is not the case... 

What do you think about adding the "role" attribute to the DimensionLink object and using that the user indicator which join link we are talking about? 

I think it is ok to leave the "role" field as an optional string since most dim links would not need it but during validating we should check if a node has unique dim links as denoted by: dim_node_id + role. Then when we convert all the dimension links to this patter we would get our dim roles "for free".

Reply:
This is a great point -- I had forgotten about the dimension roles issue that we discussed previously, and we can incorporate that nicely here. I will update to add the `role` field here.